Back-ported upstream patch to fix crash when using explain and prepared
statements with subqueries (bug #654000)


diff -Naur mysql-5.0.95.orig/sql/item_subselect.cc mysql-5.0.95/sql/item_subselect.cc
--- mysql-5.0.95.orig/sql/item_subselect.cc	2011-12-16 07:12:41.000000000 -0500
+++ mysql-5.0.95/sql/item_subselect.cc	2012-06-02 12:38:38.235237403 -0400
@@ -1821,18 +1821,23 @@
     }
     if (!select_lex->uncacheable && thd->lex->describe && 
         !(join->select_options & SELECT_DESCRIBE) && 
-        join->need_tmp && item->const_item())
+        join->need_tmp)
     {
-      /*
-        Force join->join_tmp creation, because this subquery will be replaced
-        by a simple select from the materialization temp table by optimize()
-        called by EXPLAIN and we need to preserve the initial query structure
-        so we can display it.
-       */
-      select_lex->uncacheable|= UNCACHEABLE_EXPLAIN;
-      select_lex->master_unit()->uncacheable|= UNCACHEABLE_EXPLAIN;
-      if (join->init_save_join_tab())
-        DBUG_RETURN(1);                        /* purecov: inspected */
+      item->update_used_tables();
+      
+      if (item->const_item())
+      {
+	/*
+	  Force join->join_tmp creation, because this subquery will
+	  be replaced by a simple select from the materialization
+	  temp table by optimize() called by EXPLAIN and we need to
+	  preserve the initial query structure so we can display it.
+	*/
+	select_lex->uncacheable|= UNCACHEABLE_EXPLAIN;
+	select_lex->master_unit()->uncacheable|= UNCACHEABLE_EXPLAIN;
+	if (join->init_save_join_tab())
+	  DBUG_RETURN(1);                        /* purecov: inspected */
+      }
     }
     if (item->engine_changed)
     {
